### ISSUE TITLE
snap_scheduler | always executed by a fixed node | can make it dynamic for each run?

### ISSUE THREAD

**Issue by mobilepro** [luismpcosta/snap-scheduler#1]
Thank you for sharing this to **opensource**, this is really helpful.

 

I understand the **picked = (0/1)** and **picked_by** columns controls the schedule. Just wanted to double check my understanding around this is correct and request a potential improvement.

 

**Observation**

After the first SpringBoot Node pick and execute the **snap_scheduler** entry and picked=1 set once; then the schedule is always executed by that particular node.

 

Is my observation correct?

 

**If Yes**

What if the particular node goes down for several days after running successfully for sometime? shouldn't this to be picked by some other available nodes?

I understand updating the picked=0 manually would allow some other node to pick this but ideally should you be setting picked=0 after a successful job run?

 
[Update 8th Dec 2021] - Seems If the node goes down offline after the schedule executed and set picked=1 then even the same node is not picking up the particular **snap_scheduler** entry when node comeback online.
So this needs manual intervention to retrigger the schedules.

**Comment by mobilepro** [comment]
Any update on this :-) ?

**Comment by luismpcosta** [comment]
Sorry, I forgot to answer. In about 1h I answer your questions.

Thanks!

**Comment by luismpcosta** [comment]
Hello,

Your understanding of task loading was not correct. The picked is to mark
if a task was loaded to run (true is loaded = locked to run) and picked_by
is the name of the node (only for audit). When a task time is closer to the
current date, this task is loaded to run in a node. Every load is random,
the first node that catches the task is the node where the task runs.

We have two types of task:
- One time task => this type only runs once in a specific time, we use to
send do something (e.g. send an email)   after x time (e.g. 10 minutes)  an
action occurs
-Recurring task => this is a recurring task, after running it is created a
new task (copy of executed task) to the new execution time. This approach
prevents unavailability of a node, if it occurs the task is taken by
another node to run.

I implement a failure mechanism to prevent this type of event.
- if a node takes one task and before the task execution this node was
stopped. I do not consider hard crashes, only normal stops/crashes (in the
application level). What do we do to prevent this? Before the node dies I
remove the node loaded from the database (picked and picked_by), with this
change another node catches the task and executes them. For example when
you restart a node or deploy a new application version, some node can have
tasks loaded, and this is to prevent this type of event.

We stay using this implementation in https://piscapisca.pt in two strands
*i)* guarantee that one task only runs once (with @SnapLock annotation)
*ii)* schedule task to run in a specific time with one time task (at the
moment we do not stay using recurring tasks).

This is running on production for more than one year and we didn't miss
tasks and have an audit of executed tasks (including errores when occured).


Best,
LC

**Comment by mobilepro** [comment]
Hi @luismpcosta 

Thank you for your time in patiently explaining the task loading scenarios.

It would be great if you can help to clarify the followings. And confirm if my understandings are correct.

**1.** **One time task vs Recurring task** 
      One time task: Supposed to be executed only once; eg:- Single email notification after a user action. 
      Recurring task: Recurred in a configured intervals; eg:- Daily email reports.
Is my understanding is correct?

**2.** **You mentioned** 

`-Recurring task => this is a recurring task, after running it is created a
new task (copy of executed task) to the new execution time. This approach
prevents unavailability of a node, if it occurs the task is taken by
another node to run.`

This copying of new task persisted to the table as a new entry? if I am correct it should be snap_scheduler table or snap_task_audit? 
Actually I trying to test a recurring task( every 5 minutes for example), but I am not seeing any new tasks(records in snap_scheduler )   getting created on completion of the initial tasks.
Rather I can see the "**run_at**" column getting updated with the time on which the next scheduled run should be executed along with picked = 1 and picked_by columns updated with the node which was executing the task.

Can you help to explain this part on how the new tasks copy creation done and how it's loaded to nodes ?

**3.**  **Please help to explain the expected outcomes during the following scenarios.**

**Scenario 1**: **I have setup a 1 hour recurring task and I have only **one** node**
                     Task executed by node at 1 GMT and next Run scheduled for 2 GMT. Server node goes down at 1:55 GMT for 30 minutes ( till 2:25 GMT) and node online by 2:26 GMT.
     
**Scenario 2**: **I have setup a 1 hour recurring task and I have **multiple** nodes (A,B,C,D)**         
                     Task executed by **node A** at 1 GMT and next Run scheduled for 2 GMT. Server **node A** goes down at 1:55 GMT.

**Comment by luismpcosta** [comment]
Hi,

For the first question you are correct, for the remaining questions I need
some time to review the implementation (have more than one year). Your
questions are very specific and I need to review all logic from behind.

Thanks,
LC

**Comment by mobilepro** [comment]
Thanks a lot.

**Comment by luismpcosta** [comment]
Hi,

Answers below in green style.

*1.* *One time task vs Recurring task*
One time task: Supposed to be executed only once; eg:- Single email
notification after a user action.
Recurring task: Recurred in a configured intervals; eg:- Daily email
reports.
Is my understanding is correct?

Yes, your understanding it was correct. But you can adapt to the project's
necessities.

*2.* *You mentioned*

-Recurring task => this is a recurring task, after running it is created a
new task (copy of executed task) to the new execution time. This approach
prevents unavailability of a node, if it occurs the task is taken by
another node to run.

This copying of new task persisted to the table as a new entry? if I am
correct it should be snap_scheduler table or snap_task_audit?
Actually I trying to test a recurring task( every 5 minutes for example),
but I am not seeing any new tasks(records in snap_scheduler ) getting
created on completion of the initial tasks.
Rather I can see the "*run_at*" column getting updated with the time on
which the next scheduled run should be executed along with picked = 1 and
picked_by columns updated with the node which was executing the task.

Can you help to explain this part on how the new tasks copy creation done
and how it's loaded to nodes ?

All task are created on snap_scheduler table and create one entry in
snap_task_audit for each run, when is a recurring task the runa_at column
it is updated to next run time. If the recurring time is less than 2x
database polling interval the task was loaded to run in the current
instance, if it is greater then it was loaded by default load tasks.

*3.* *Please help to explain the expected outcomes during the following
scenarios.*

*Scenario 1*: *I have setup a 1 hour recurring task and I have only one
node*
Task executed by node at 1 GMT and next Run scheduled for 2 GMT. Server
node goes down at 1:55 GMT for 30 minutes ( till 2:25 GMT) and node online
by 2:26 GMT.

This task does not run and is never loaded to run again. The only way to
run it again is to change the run_at column in the DB to a time greater
than the current date.

*Scenario 2*: *I have setup a 1 hour recurring task and I have multiple
nodes (A,B,C,D)*
Task executed by *node A* at 1 GMT and next Run scheduled for 2 GMT.
Server *node
A* goes down at 1:55 GMT

If one instance is available at 2GMT to run the task, the task is loaded
and executed in one of the available instances.

**Comment by luismpcosta** [comment]
Hi,

Answers below in green style.

*1.* *One time task vs Recurring task*
One time task: Supposed to be executed only once; eg:- Single email
notification after a user action.
Recurring task: Recurred in a configured intervals; eg:- Daily email
reports.
Is my understanding is correct?

Yes, your understanding it was correct. But you can adapt to the project's
necessities.

*2.* *You mentioned*

-Recurring task => this is a recurring task, after running it is created a
new task (copy of executed task) to the new execution time. This approach
prevents unavailability of a node, if it occurs the task is taken by
another node to run.

This copying of new task persisted to the table as a new entry? if I am
correct it should be snap_scheduler table or snap_task_audit?
Actually I trying to test a recurring task( every 5 minutes for example),
but I am not seeing any new tasks(records in snap_scheduler ) getting
created on completion of the initial tasks.
Rather I can see the "*run_at*" column getting updated with the time on
which the next scheduled run should be executed along with picked = 1 and
picked_by columns updated with the node which was executing the task.

Can you help to explain this part on how the new tasks copy creation done
and how it's loaded to nodes ?

All task are created on snap_scheduler table and create one entry in
snap_task_audit for each run, when is a recurring task the runa_at column
it is updated to next run time. If the recurring time is less than 2x
database polling interval the task was loaded to run in the current
instance, if it is greater then it was loaded by default load tasks.

*3.*  *Please help to explain the expected outcomes during the following
scenarios.*

*Scenario 1*: *I have setup a 1 hour recurring task and I have
only one node*
Task executed by node at 1 GMT and next Run scheduled for 2 GMT. Server
node goes down at 1:55 GMT for 30 minutes ( till 2:25 GMT) and node online
by 2:26 GMT.

This task does not run and is never loaded to run again. The only way to
run it again is to change the run_at column in the DB to a time greater
than the current date.

*Scenario 2*: *I have setup a 1 hour recurring task and I
have multiple nodes (A,B,C,D)*
Task executed by *node A* at 1 GMT and next Run scheduled for 2 GMT.
Server *node
A* goes down at 1:55 GMT

If one instance is available at 2GMT to run the task, the task is loaded
and executed in one of the available instances.

Regards,
LC

**Comment by mobilepro** [comment]
Hi @luismpcosta ,

Thank you for the clarifications for all the queries. That was really helpful to get clarify.

So putting everything together I see there is an issue when all nodes are down before next scheduled run trigger, where we need a manual intervention to update the "runAt" column so that once the nodes are up it will pool and execute the task.

But ideally there should not be any manual intervention required as there is a possibility to programmatically detect that a particular schedule is skipped { sample logic such as .....(**currentTime during pooling**) **>** **(Time as in lastest RunAT Column) + (interval Between Runs)** }

Please share your thoughts on this ?

**Comment by luismpcosta** [comment]
Hi,

Answers below in green style.

*1.* *One time task vs Recurring task*
One time task: Supposed to be executed only once; eg:- Single email
notification after a user action.
Recurring task: Recurred in a configured intervals; eg:- Daily email
reports.
Is my understanding is correct?

Yes, your understanding it was correct. But you can adapt to the project's
necessities.

*2.* *You mentioned*

-Recurring task => this is a recurring task, after running it is created a
new task (copy of executed task) to the new execution time. This approach
prevents unavailability of a node, if it occurs the task is taken by
another node to run.

This copying of new task persisted to the table as a new entry? if I am
correct it should be snap_scheduler table or snap_task_audit?
Actually I trying to test a recurring task( every 5 minutes for example),
but I am not seeing any new tasks(records in snap_scheduler ) getting
created on completion of the initial tasks.
Rather I can see the "*run_at*" column getting updated with the time on
which the next scheduled run should be executed along with picked = 1 and
picked_by columns updated with the node which was executing the task.

Can you help to explain this part on how the new tasks copy creation done
and how it's loaded to nodes ?

All task are created on snap_scheduler table and create one entry in
snap_task_audit for each run, when is a recurring task the runa_at column
it is updated to next run time. If the recurring time is less than 2x
database polling interval the task was loaded to run in the current
instance, if it is greater then it was loaded by default load tasks.

*3.*  *Please help to explain the expected outcomes during the following
scenarios.*

*Scenario 1*: *I have setup a 1 hour recurring task and I have
only one node*
Task executed by node at 1 GMT and next Run scheduled for 2 GMT. Server
node goes down at 1:55 GMT for 30 minutes ( till 2:25 GMT) and node online
by 2:26 GMT.

This task does not run and is never loaded to run again. The only way to
run it again is to change the run_at column in the DB to a time greater
than the current date.

*Scenario 2*: *I have setup a 1 hour recurring task and I
have multiple nodes (A,B,C,D)*
Task executed by *node A* at 1 GMT and next Run scheduled for 2 GMT.
Server *node
A* goes down at 1:55 GMT

If one instance is available at 2GMT to run the task, the task is loaded
and executed in one of the available instances.

Regards,
LC

**Comment by mobilepro** [comment]
Please ignore my last question, i will test more around this.

Based on the tests so far I see the following behavior which looks great.

1. When the RunAt time is near by and picked=0 then the pooling will assign the node to the task by setting picked_by and picked=1.
2. Just after or exactly at RunAT time ( depending on pooling interval) the respective node execute the Job.
3. After the job execution picked_by set to EMPTY/Null and picked set to picked=0 so that the JOB is free to be executed by any node.

**Comment by luismpcosta** [comment]
Hi,

Yes, all that you describe in your email are correct. If you need some
additional information tell me.


Regards,
LC
